### PR TITLE
rm-old-base: remove ifdefs for pre-4.13 bases

### PR DESCRIPTION
**Pull request by NadiaYvette**
**Template B: This PR does not modify behaviour or interface**

4.13 and older have fallen out of the support window. Hence this commit removes code only conditionally included for base 4.13 and older. Some occasional transitive removals were implied and done in this same commit.

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [ ] Is this a PR that fixes CI? If so, it will need to be backported to older cabal release branches (ask maintainers for directions).

### CODE REVIEW

**Comment by geekosaur** [comment]
You're supposed to pick a template based on whether your PR changes the user-visible or programming API. This one should not, so you would use template B.

**Review by andreabedini — APPROVED** [review]
LGTM. Thank you so much! <3

File: Cabal-syntax/src/Distribution/Compat/Prelude.hs [hunk 1]
```diff
@@ -243,11 +234,11 @@ import Data.Ord (comparing)
 import Data.Proxy (Proxy (..))
 import Data.Set (Set)
 import Data.String (IsString (..))
+import Data.Typeable (TypeRep, Typeable, typeRep)
 import Data.Void (Void, absurd, vacuous)
 import Data.Word (Word, Word16, Word32, Word64, Word8)
 import Distribution.Compat.Binary (Binary (..))
 import Distribution.Compat.Semigroup (Semigroup (..), gmappend, gmempty)
-import Distribution.Compat.Typeable (TypeRep, Typeable, typeRep)
```
Comment (fgaz) [inline Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: Can we remove the module?

File: Cabal-syntax/src/Distribution/Fields/ParseResult.hs [hunk 2]
```diff
@@ -18,22 +18,21 @@ module Distribution.Fields.ParseResult
...
+import Prelude
+  ( Applicative (..)
+  , Either (..)
+  , Functor (..)
+  , Maybe (..)
+  , Monad (..)
+  , String
+  , ($)
+  , (++)
+  )
```
Comment (fgaz) [inline Cabal-syntax/src/Distribution/Fields/ParseResult.hs:35]: Can we just import Distribution.Compat.Prelude?

File: Cabal/src/Distribution/Simple/Build/PathsModule/Z.hs [hunk 3]
```diff
@@ -85,11 +85,7 @@ render z_root = execWriter $ do
   then do
     tell "#if defined(VERSION_base)\n"
     tell "\n"
-    tell "#if MIN_VERSION_base(4,0,0)\n"
     tell "catchIO :: IO a -> (Exception.IOException -> IO a) -> IO a\n"
-    tell "#else\n"
-    tell "catchIO :: IO a -> (Exception.Exception -> IO a) -> IO a\n"
-    tell "#endif\n"
```
Comment (fgaz) [inline Cabal/src/Distribution/Simple/Build/PathsModule/Z.hs:92]: This must be restored. While we don't support building Cabal itself with old GHCs, iirc we do support building packages with new Cabal and old GHCs. Paths module generation affects the latter.

File: Cabal-syntax/Cabal-syntax.cabal
Comment (fgaz) [inline Cabal-syntax/Cabal-syntax.cabal:1]: The change must be reflected in the build-depends field

File: Cabal-tests/Cabal-tests.cabal
Comment (fgaz) [inline Cabal-tests/Cabal-tests.cabal:1]: The change must be reflected in the build-depends field

File: Cabal/Cabal.cabal
Comment (fgaz) [inline Cabal/Cabal.cabal:1]: The change must be reflected in the build-depends field

(also in cabal-install)

File: cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs
Comment (fgaz) [inline cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs:1]: Why was this test removed?

**Review by fgaz — CHANGES_REQUESTED** [review]

File: cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs
Comment (andreabedini) [reply cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs:1]: D'oh, sorry my bad. I misread the condition in the testcase. I'll restore it.

Comment (andreabedini) [reply cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs:1]: Done

**Comment by andreabedini** [comment]
@fgaz Can you have another look when you have time?

**Comment by andreabedini** [comment]
@ulysses4ever @Mikolaj @Kleidukos could you give a review?

**Comment by NadiaYvette** [comment]
> @fgaz @ulysses4ever @Mikolaj @Kleidukos could you give a review?

Thank you so much Andrea! That's a lot of LOC you put into the fixups!

File: Cabal/src/Distribution/Compat/Async.hs [hunk 4]
```diff
@@ -150,12 +148,10 @@ data AsyncCancelled = AsyncCancelled
 
 {- FOURMOLU_DISABLE -}
```
Comment (geekosaur) [inline Cabal/src/Distribution/Compat/Async.hs:149]: I think you can remove the Fourmolu directives when you're getting rid of the CPP.

File: cabal-install/src/Distribution/Deprecated/ParseUtils.hs [hunk 5]
```diff
@@ -128,11 +128,6 @@ instance Monad ParseResult where
   ParseOk ws x >>= f = case f x of
```
Comment (geekosaur) [inline cabal-install/src/Distribution/Deprecated/ParseUtils.hs:128]: Same here (the disable directive is several lines before the diff hunk).

**Review by geekosaur — APPROVED** [review]
It might be worth looking for more instances of `FOURMOLU_DISABLE` and see if CPP has been removed; in general, CPP causes fourmolu to choke.

File: Cabal/src/Distribution/Compat/Async.hs (hunk 4 above)
Comment (NadiaYvette) [reply Cabal/src/Distribution/Compat/Async.hs:149]: I can do that.  Now pushed!

File: cabal-install/src/Distribution/Deprecated/ParseUtils.hs (hunk 5 above)
Comment (NadiaYvette) [reply cabal-install/src/Distribution/Deprecated/ParseUtils.hs:128]: I bundled it with the change for the above & used --amend to fold it into the last commit.

**Comment by Kleidukos** [comment]
Yes! Less CPP make happier tooling

**Comment by andreabedini** [comment]
Maybe something went wrong with the rebase I had changed the bounds in `Cabal.cabal` in my commit acffcd425c2ec7064495550be8f026fac0bc76a6

File: Cabal-syntax/src/Distribution/Compat/Prelude.hs (hunk 1 above)
Comment (fgaz) [reply Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: This was marked as resolved, but the module is still there. Can we not remove it?

File: templates/Paths_pkg.template.hs [hunk 6]
```diff
@@ -38,11 +38,7 @@ import System.Environment (getExecutablePath)
 {% if supportsCpp %}
 #if defined(VERSION_base)
 
-#if MIN_VERSION_base(4,0,0)
 catchIO :: IO a -> (Exception.IOException -> IO a) -> IO a
-#else
-catchIO :: IO a -> (Exception.Exception -> IO a) -> IO a
-#endif
```
Comment (fgaz) [inline templates/Paths_pkg.template.hs:42]: I think this should also be restored, similar to https://github.com/haskell/cabal/pull/10092#discussion_r1639858400. iirc the Z module is generated from this file.

**Review by fgaz — CHANGES_REQUESTED** [review]
Some of my previous comments are still unresolved, I marked them as such.

**Comment by Mikolaj** [comment]
> Maybe something went wrong with the rebase I had changed the bounds the cabal files in my commit [acffcd4](https://github.com/haskell/cabal/commit/acffcd425c2ec7064495550be8f026fac0bc76a6).
>
> Edit: I pushed my last commit ([9ce4213](https://github.com/haskell/cabal/commit/9ce421302920bdbbee8a6e1a51f866328c9ea7d1)) to https://github.com/andreabedini/cabal/tree/rm-old-base-001

Right here it's as it should be. That's rather disturbing. Was this mergify's rebase or your manual one? Can you recover from this corruption?

**Comment by andreabedini** [comment]
@fgaz @Mikolaj I restored the lower bounds on base.

**Comment by ulysses4ever** [comment]
@fgaz you requested changes in this PR, and the issue, they claim, was fixed (I didn't check). Could you see if you could reevaluate it and maybe approve?

**Comment by geekosaur** [comment]
FWIW everything except the `import` is marked `Outdated`, suggesting they've already been fixed. I'm validating locally with the `import` removed.

File: templates/Paths_pkg.template.hs (hunk 6 above)
Comment (ulysses4ever) [reply templates/Paths_pkg.template.hs]: @andreabedini any thoughts? It hasn't been restored yet, am I right?

File: Cabal-syntax/src/Distribution/Compat/Prelude.hs (hunk 1 above)
Comment (geekosaur) [reply Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: The build fails without it, FWIW.

**Comment by geekosaur** [comment]
I stand corrected; I misunderstood what was being requested, and removed the wrong thing. Testing again now.

File: Cabal-syntax/src/Distribution/Compat/Prelude.hs (hunk 1 above)
Comment (fgaz) [reply Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: It builds on my machine and I didn't find any other reference to that module using `git grep`. What error are you getting?

**Comment by fgaz** [comment]
> @fgaz you requested changes in this PR, and the issue, they claim, was fixed (I didn't check). Could you see if you could reevaluate it and maybe approve?

All opened conversations are still unresolved

**Comment by geekosaur** [comment]
```
[27 of 51] Compiling Distribution.Solver.Types.ConstraintSource ( src/Distribution/Solver/Types/ConstraintSource.hs, /home/allbery/Sources/cabal/dist-newstyle/build/x86_64-linux/ghc-8.8.4/cabal-install-solver-3.13.0.0/build/Distribution/Solver/Types/ConstraintSource.o )

src/Distribution/Solver/Types/ConstraintSource.hs:48:3: error:
    parse error on input ‘-- | The source of the constraint is not specified.’
   |
48 |   -- | The source of the constraint is not specified.
   |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: Cabal-syntax/src/Distribution/Compat/Prelude.hs (hunk 1 above)
Comment (ulysses4ever) [reply Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: I wonder how soon 8.4.4 will fall out of our support window.

Comment (geekosaur) [reply Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: Looks like the error was introduced by #9900? (And may be specific to old ghcs)

Comment (geekosaur) [reply Cabal-syntax/src/Distribution/Compat/Prelude.hs:250]: Just filed #10226; when this is rebased on that, it should fix the Haddock error.

File: templates/Paths_pkg.template.hs (hunk 6 above)
Comment (geekosaur) [reply templates/Paths_pkg.template.hs]: Looks restored to me (that is, this no longer shows up in the diff).

File: cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs
Comment (geekosaur) [reply cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs:1]: If you restored it, you didn't push it; it's still showing as deleted.

Comment (geekosaur) [reply cabal-testsuite/PackageTests/Regression/T6906/cabal.test.hs:1]: I just restored it from `master`.

**Comment by geekosaur** [comment]
@fgaz, I think we've dealt with the remaining issues, could you confirm?

**Comment by andreabedini** [comment]
@geekosaur @fgaz @ulysses4ever Thank you for the help. I am afraid some changes kept getting lost in the rebases :-/

**Review by fgaz — APPROVED** [review]
Looks good!

**Pull request merged**